README.md
<div align="center"><h1>Dispage</h1></div>

## About

dispage is a powerful [Node.js](https://nodejs.org) module that allows you to easily create button to switch over embed pages on a discord message.
- [discord.js](https://npmjs.com/package/discord.js)
- [Discord API](https://discord.com/developers/docs/intro).

## Dependencies

**Needs discord.js v14.6.0 or higher.**
If discord.js isn't installed yet :
```sh-session
npm install discord.js@latest
```

## Installation
```sh-session
npm i dispage
```

## Importing
### Javascript
```js
const Dispage = require('dispage')
```
### Typescript
```ts
import Dispage from 'dispage';
```
### Simple example
```js
const embeds = [
    new MessageEmbed().setDescription('Embed #1'),
    new MessageEmbed().setDescription('Embed #2')
];
return new Dispage()
    .setEmbeds(embeds)
    .start(message)
```
![Simple example](https://iili.io/Xf3MhP.png)

### If you want to push things further...
```js
const embeds = [
    'embed 1',
    'embed 2 !!',
    'embed 3 !?!',
    'embed 4 !!!!!!!!?'
].map((desc, i) => new Discord.EmbedBuilder()
    .setTitle(`📃 Page #${i + 1}`)
    .setColor('BLURPLE')
    .setDescription(desc))

new Dispage()
    .setEmbeds(embeds)
    // This style will be given to every button that doesn't have a specific style
    .setMainStyle('Secondary')
    .editButton('previous', { emoji: "824240081409540157", label: "Previous" })
    // Setting a property (here, style) to null will remove it. 👇
    .editButton('stop', { emoji: "961338862259544114", label: null })
    .editButton('next', { emoji: "824240024509874186", label: "Next" })
    // We can edit as many buttons as we want and as many times as we wish
    .editButton('stop', { style : "Success" })
    // You can create your own custom button
    .addButton({
        label: "Go to Message",
        style: "Link",
        url: "https://discord.com/channels/937626764916719626/937633296878293002/978956060972974101"
    })
    // Finally after having setup everything. We can start
    .start(message)
```

![Page 1](https://iili.io/XnAgvR.png)
![Page 2](https://iili.io/XnAUYv.png)

## Properties
> Instead of (for example) setting `index` to 1 by hand. Use the build-in (`setIndex`) method. Same for `embeds` (.setEmbeds), `ended` (.end()), duration (.setDuration) etc..

| **Property**  | **Type**                                  | **Default Value**        |
| ------------- | ----------------------------------------- | ------------------------ |
| `client`      | `Client`                                  | `null`                   |
| `index`       | `number`                                  | `0`                      |
| `embeds`      | `Embed[]`                                 | `[]`                     |
| `message`     | `Message`                                 | `null`                   |
| `interaction` | `Interaction`                             | `null`                   |
| `collector`   | `InteractionCollector<ButtonInteraction>` | `null`                   |
| `reply`       | `Message`                                 | `null`                   |
| `ended`       | `boolean`                                 | `false`                  |
| `started`     | `boolean`                                 | `false`                  |
| `deleted`     | `boolean`                                 | `false`                  |
| `duration`    | `number`                                  | `60000`                  |
| `mainStyle`   | `MessageButtonStyle`                      | `"PRIMARY"`              |
| `buttons`     | `MessageButtonOptions[]`                  | `MessageButtonOptions[]` |

## Methods
> On the `Return` column, `this` means that the methode returns the original instance of the class. Like discord.js's MessageEmbed where we can call multiple times multiple methods following themselves. Like this: `new Dispage().setMainStyle('SECONDARY').setIndex(2)` etc..

**__*PS*__: ⚠ MEANS __[DEPRECATED]__ (Meaning you shouldn't use it anymore)**

|       **Name**        |         **Arguments**          |         **Return**         |
| :-------------------: | :----------------------------: | :------------------------: |
|    `setMainStyle`     |      `MessageButtonStyle`      |           `this`           |
| `showDisabledButtons` |           `boolean`            |           `this`           |
|     `removeUser`      |             `User`             |           `this`           |
|       `addUser`       |             `User`             |           `this`           |
|       `setUser`       |             `User`             |           `this`           |
|      `addButton`      |     `MessageButtonOptions`     |           `this`           |
|    `removeButton`     |            `string`            |           `this`           |
|     `editButton`      | `string, MessageButtonOptions` |           `this`           |
|       `getRows`       |           `boolean`            |    `MessageActionRow[]`    |
|      `setEmbeds`      |           `Embed[]`            |           `this`           |
|      `addEmbed`       |            `Embed`             |           `this`           |
|      `addEmbeds`      |           `Embed[]`            |           `this`           |
|     `_fixEmbeds`      | `Embed[] / Embed / Embed[][]`  |         `Embed[]`          |
|     `setDuration`     |            `number`            |           `this`           |
|     `addDuration`     |            `number`            |           `this`           |
|        `next`         |               ❌                |      `Promise<this>`       |
|      `previous`       |               ❌                |      `Promise<this>`       |
|    `changeToPage`     |            `number`            |     `Promise<Message>`     |
|   `doesIndexExist`    |            `number`            |         `boolean`          |
|      `setIndex`       |            `Number`            |           `this`           |
|        `edit`         |      `MessageEditOptions`      | `Promise<Discord.Message>` |
|   `disableButtons`    |               ❌                |           `void`           |
|         `end`         |               ❌                |           `void`           |
|       `delete`        |               ❌                |           `void`           |
|       `update`        |               ❌                |           `void`           |
|       `getOpts`       |           `Boolean`            |      `MessageOptions`      |
|      `isMessage`      |               ❌                |         `boolean`          |
|    `isInteraction`    |               ❌                |         `boolean`          |
|       `canEdit`       |               ❌                |         `boolean`          |
|     `isValidCtx`      |           `Context`            |         `boolean`          |
|   `checkForErrors`    |           `Context`            |         `string[]`         |
|        `start`        |           `Context`            |      `Promise<this>`       |

## Trouble
Having a problem using <a href="https://npmjs.com/package/dispage">**dispage**</a> ? Open an <a href="https://github.com/voxlinou1/dispage/issues">issue</a> on <a href="https://github.com/voxlinou1/dispage">Github<a> & Don't hesitate to send a message to the discord tag below 👇 !

## Credits
Made by voxlinou1 (*`Vox#6198`* on discord)


CHANGES.md
# 1.0.0
- Published the package

# 1.0.1
- Better README.md
- Added typescript notation support for typescript programmers

# 1.0.2
- Added a CHANGES.md (current file)
- Typescript declaration files are now automatic using `npx tsc index.js -d --allowJs`
- Fixed a bug with "instanceof" not working locally (not knowing if a message is a discord.js Message, embeds is MessageEmbed, interaction is CommandInteraction etc..).
- Transformed some static properties into getters for more logic and readability.
- next() and previous() either return the instance when using just after the constructor or returns a promise when using after the .start() promise.
- Possibility to control buttons with .addButton, .removeButton & .editButton

# 1.0.3
- Added .addUser, .removeUser and .setUser to add a user id to the people who can interact with the buttons.
- [DEPRECATED] .addUserID by .addUser for better efficiency and flexibility.
- Added .setMainStyle() to change from the default PRIMARY style (for left and right mostly.)
- Users not allowed in the embed are now getting their interaction deferred and given an ephemeral response
- Made some properties & methods private for better experience with IDE's docs
- .getRow() replaced now by .getRow**s**();
- Improved index.d.ts.

# 1.0.4
- Removed automatic footer as it's not user-friendly
- Collected moved to another method for possibility of change
- If the stop button is pressed, the embeds will disapear but if it's only the time that runs out. The buttons will be disabled only.
- Built-in .doesIndexExist()
- Types for typescript (Context, User, Message and Embed & ButtonId)
- Methods that starts with _ should only be used inside the class.
- Added an example.js file to copy/paste
- Added a tsconfig.json for efficiency and reliable experience while updating the package
- .addDuration not crashing when .setDuration has been used with a none-number
- async every method for better error handling and crash path tracability
- .canEdit that returns true if dispage has neither stopped, been deleted, the time has run out and the page system has been sent and successfully started
- .editButton, .createButton and .removeButton to manipulate buttons on screen and create existing imaginations
- Better JSDoc for none-typescript cases.
# 1.0.5
- Fixed a small bug that crashed when the embed page system gets deleted.
# 1.0.6
- Made it compatible for v14
- Fixed bug that made the buttons still enabled after dispage turning off
- Fixed bug that crashed the process when the message is deleted while the buttons are still active
- Fixed example.js including intents, builders and button style
- Fixed buttons that still appear even if .showDisabledButtons is disabled

index.js
const {
    Client,
    Message,
    EmbedBuilder,
    ActionRowBuilder,
    ButtonBuilder,
    MessageEditOptions,
    MessageOptions,
    Interaction,
    InteractionCollector,
    ButtonStyle,
    MessageType,
    ComponentType
} = require('discord.js');

module.exports = class Dispage {
    /** @param {Client} client  */
    constructor(client) {
        this.client = client;
        this.stopReason = null;
        this.index = 0;
        /** @type {EmbedBuilder[]} */
        this.embeds = [];
        this.message = null;
        this.interaction = null;
        this.collector = null;
        this.reply = null;
        this.ended = false;
        this.stopReason = null;
        this.started = false;
        this.deleted = false;
        this.duration = 1000 * 60;
        this.mainStyle = "Primary";
        this.styles = Object.keys(ButtonStyle);
        this.buttons = [
            ['previous', "⬅️"],
            ['stop', "⏹️", "DANGER"],
            ['next', "➡️"]
        ].map(arr => {
            return {
                customId: arr[0],
                emoji: arr[1],
                style: arr[2],
                label: arr[3]
            }
        });
        this.filter = (int) => this.users.includes(int.user.id);
        this._hdb = false;
        /** @type {Set<string>} */
        this._userIDs = new Set();
    }
    get users() {
        return Array.from(this._userIDs)
    }
    get ctx() {
        return this.message ? "MESSAGE" : this.interaction ? "INTERACTION" : null;
    }
    get type() {
        return this.ctx?.toLowerCase() ?? null;
    }
    get endUntill() {
        return this.collector ? (Date.now() - this.collector.options.time) : null;
    }
    /** @returns {EmbedBuilder} */
    get currentEmbed() {
        return this.embeds[this.index];
    }
    _getId(user) {
        return this.client?.users.resolve(user)?.id || user;
    }
    /**
     * Sets the main style for buttons who don't have a per-defined style.
     * @param {ButtonBuilderStyle} style 
     * @returns {this}
     */
    setMainStyle(style) {
        this.mainStyle = style;
        return this;
    }
    /**
     * If buttons that aren't available should be hidden.
     * @param {boolean} should 
     * @returns {this}
     */
    hideDisabledButtons(should = false) {
        this._hdb = should;
        return this;
    }
    /**
     * Rmoves a user from the list of users allowed to interact with the buttons
     * @param {UserResolvable} user 
     * @returns {this}
     */
    removeUser(user) {
        this._userIDs.delete(this._getId(user));
        return this;
    }
    /**
     * Adds a user to the list of users allowed to interact with the buttons
     * @param {UserResolvable} user 
     * @returns {this}
     */
    addUser(user) {
        this._userIDs.add(this._getId(user));
        return this;
    }
    /**
     * Sets the user being the only one allowed to interact with the buttons
     * @tip To have multiple users beeing able to interact with the buttons, use .addUser()
     * @param {UserResolvable} user 
     * @returns {this}
     */
    setUser(user) {
        this._userIDs = new Set([this._getId(user)]);
        return this;
    }
    /**
     * Creates a button
     * @param {object} o
     * @returns {this}
     */
    addButton(o) {
        this.buttons.push(o);
        return this;
    }
    /**
     * Removes every existing button
     * @returns {this}
     */
    removeAllButtons() {
        this.buttons = [];
        return this;
    }
    /**
     * Find a button with a function to remove it.
     * @param {Function} query 
     * @returns {this}
     */
    findButtonAndRemove(query) {
        const i = this.buttons.findIndex(query);
        if (i !== -1) this.buttons.splice(i, 1)
        return this;
    }
    /**
     * Removes a specific button with it's id
     * @param {string} customId
     * @returns {this}
     */
    removeButton(customId) {
        const i = this.buttons.findIndex(btn => btn?.customId === customId);
        if (i !== -1) this.buttons.splice(i, 1)
        return this;
    }
    /**
     * Edit an existing button with it's id
     * @param {string} customId
     * @param {object} o
     * @returns {this}
     */
    editButton(customId, o) {
        const i = this.buttons.findIndex(btn => btn.customId === customId);
        if (i < 0) this.addButton(o)
        else {
            this.buttons[i] = { ...this.buttons[i], ...o };
        }
        return this;
    }
    /**
     * Gets the current button component row
     * @param {boolean} disabled 
     * @returns {ActionRowBuilder | null}
     */
    getRows(disabled = false) {
        let ar = new ActionRowBuilder()
        let previous = this.embeds[this.index - 1];
        let next = this.embeds[this.index + 1];
        let format = a => {
            let d = disabled
                || (a.customId == "previous" && !previous)
                || (a.customId == "next" && !next)
                || this.ended
                || false;

            if (this._hdb && d) return;

            let button = new ButtonBuilder()
                .setDisabled(d)
                .setStyle(a.style || this.mainStyle)
            if (a.customId) button.setCustomId(a.customId)
            if (a.emoji) button.setEmoji(a.emoji)
            if (a.label) button.setLabel(a.label)
            if (a.url) button.setURL(a.url)
            return ar.addComponents(button);
        }
        this.buttons.forEach(format)
        if (this._hdb) ar.components = ar.components.filter(c =>!c.disabled );
        return ar.components.length ? [ar] : [];
    }
    /**
     * Sets the embed array.
     * @param {EmbedBuilder[]} embeds 
     * @returns {this}
     * @example
     * .setEmbeds([
     *     new EmbedBuilder().setDescription('embed 1'),
     *     new EmbedBuilder().setDescription('embed 2!'),
     *     new EmbedBuilder().setDescription('embed 3!!')
     * ])
     */
    setEmbeds(embeds) {
        embeds = this._fixEmbeds(embeds);
        this.embeds = embeds;
        return this;
    }

    /**
     * Add an embed to the array of embed
     * @param {EmbedBuilder} embed
     * @returns {this}
     * @example
     * .addEmbed(new EmbedBuilder().setDescription('embed 4??'))
     */
    addEmbed(embed) {
        let embeds = this._fixEmbeds(embed)
        return this.setEmbeds([...this.embeds, ...embeds]);
    }

    /**
     * Add multiple arrays to the array of embeds
     * @param {EmbedBuilder[]} embeds
     * @returns {this}
     * @example
     * .addEmbeds([
     *   new EmbedBuilder().setDescription('embed 5!!'),
     *   new EmbedBuilder().setDescription('embed 6!!'),
     *   new EmbedBuilder().setDescription('embed 7!!'),
     * ])
     */
    addEmbeds(embeds) {
        embeds = this._fixEmbeds(embeds);
        embeds.forEach(embed => this.addEmbed(embed));
        return this;
    }

    /**
     * Returns an array of embeds whether the parameter is ONE embed, an array of embeds or an array of arrays
     * @param {EmbedBuilder | EmbedBuilder[] | EmbedBuilder[][]} embeds 
     * @returns {EmbedBuilder[]}
     */
    _fixEmbeds(embeds) {
        embeds = Array.isArray(embeds[0]) ? embeds.flat(1) : embeds
        embeds = embeds?.type ? [embeds] : embeds;
        return embeds;
    }

    /**
     * Sets the number of milliseconds before the embed page system stops
     * @param {number} duration 
     * @returns {this}
     */
    setDuration(duration) {
        this.duration = duration
        return this;
    }

    /**
     * Adds milliseconds to the current duration of the page system
     * @param {number} duration 
     * @returns {this}
     */
    addDuration(duration) {
        this.duration = (this.duration ?? 0) + duration;
        return this;
    }

    /**
     * Go to the next page
     * @returns {Promise<Message | false>}
     */
    next(int) {
        let newIndex = this.index + 1
        if (this.canEdit() && this.doesIndexExist(newIndex))
            return this.changeToPage(newIndex, int)
        return Promise.resolve(false);
    }

    /**
     * Get to the previous page
     * @returns {Promise<Message | false>}
     */
    previous(int) {
        let newIndex = this.index - 1
        if (this.canEdit() && this.doesIndexExist(newIndex))
            return this.changeToPage(newIndex, int)
        return Promise.resolve(false);
    }
    /**
     * Checks if the index exists on the embed array
     * @param {number} index 
     * @returns {boolean}
     */
    doesIndexExist(index) {
        return Boolean(this.embeds[index]);
    }
    /**
     * Change pages
     * @param {number} index
     * @returns {Promise<Message>} 
     */
    async changeToPage(index, int) {
        this.setIndex(index)
        return await int.update(this.getOpts())
    }
    /**
     * @param {number} index 
     * @returns {this}
     */
    setIndex(index) {
        this.index = index;
        return this;
    }

    /**
     * End the embed page system
     * @param {string} reason 
     * @returns {Promise<void | Message>}
     */
    async end(reason) {
        this.stopReason = reason;
        this.ended = true;
        if (!this.canEdit()) return await Promise.resolve();
        return await this.reply.edit(this.getOpts(true));
    }

    /**
     * Delete the embed page system message
     * @returns {Promise<Message>}
     */
    async delete() {
        this.deleted = true;
        this.end("messageDelete")
        if (this.reply) return await this.reply.delete();
        else return await Promise.reject('No message to edit.');
    }

    /**
     * @param {boolean} disabled 
     * @returns {MessageOptions}
     */
    getOpts(disabled = false) {
        return {
            embeds: [this.currentEmbed],
            components: this.getRows(disabled)
        };
    }
    isMessage() {
        return this.ctx ? this.ctx === "MESSAGE" : null;
    }
    isInteraction() {
        return this.ctx ? this.ctx === "INTERACTION" : null;
    }
    canEdit() {
        return this.ctx && this.started && !this.deleted;
    }
    /**
     * Check if the argument can be used in the .start() method
     * @param {Message | Interaction} ctx 
     * @returns {boolean}
     */
    isValidCtx(ctx) {
        return ctx &&
            (ctx.isRepliable() || [
                MessageType.Default, MessageType.Reply,
                MessageType.ChatInputCommand, MessageType.ContextMenuCommand
            ].includes(ctx.type))
    }
    /**
     * Creates a collector of buttons of the message with the duration
     * @param {Message} reply 
     * @returns {InteractionCollector}
     */
    _createCollector(reply) {
        return reply.createMessageComponentCollector({
            componentType: ComponentType.Button,
            time: this.duration,
        });
    }
    /**
     * Check if any problems are existing in the current 
     * @param {Message | Interaction} msg 
     * @returns {string[]}
     */
    checkForErrors(ctx) {
        let errors = []
        function add(error) { errors.push(error) };
        if (this.ended) add('Dispage has ended already.')
        if (this.deleted) add('Dispage has previously been deleted.')
        if (!ctx) add('No message/interaction given.');
        else {
            if (ctx.author) {
                this.message = ctx;
                this.addUser(ctx.author.id)
            } else if (ctx.user) {
                this.interaction = ctx;
                this.addUser(ctx.user.id)
            } else add(`.start(<ctx>) -> ctx is neiher an interaction nor a message.`);
        }
        if (!this.embeds.length) add('Embed array is empty.')
        else if (!this.embeds[this.index]) add(`No embed at index ${this.index}`);

        if (typeof (this.index) !== "number"
            || isNaN(this.index)
            || this.index < 0) add(`Invalid index ${this.index}`);

        if (typeof (this.filter) !== "function")
            add(`.filter should be a function but is a ${typeof (this.filter)}.`)
        if (typeof (this._hdb) !== "boolean")
            add(`.hideDisabledButtons() should be a boolean but is a ${typeof (this._sdb)}`)
        if (typeof (this.duration) !== "number")
            add(`.duration should be a number but is a ${typeof this.duration}`)

        if (!this.styles.includes(this.mainStyle)) add('.mainStyle is not a valid button style')
        if (!Array.isArray(this.buttons)) add('.buttons is not an array as expected.')

        this.users.forEach((id, i) => {
            if (typeof (id) !== "string") add(`ID "${id}" at index ${i} is not a string but a ${typeof (id)}.`)
            else if (id.split(' ').some(c => !Number(c))) add(`ID "${id}" at index ${i} is not a valid user ID.`);
        })
        return errors;
    }
    /**
     * Execute the embed page system 
     * @param {Message | Interaction} ctx 
     * @returns {Promise<this>}
     */
    async start(ctx) {
        let errs = this.checkForErrors(ctx)
        const err = new Error(`Errors ecountered :\n${errs.map((e, i) => `${i} - ${e}`).join('\n')}`);
        if (errs.length) throw err;
        if (this.started) throw new Error('Dispage already started')
        if (!this.users.length) this._userIDs.add(ctx.author?.id || ctx.user?.id);

        /** @type {Message} */
        this.reply = await this[this.type].reply({
            fetchReply: true,
            ...this.getOpts()
        }).then(res => {
            this.started = true;
            return res;
        })

        if (!this.reply) throw new Error('Could not fetch reply of the message/interaction');
        if (!this.getRows()) return;

        this.collector = this._createCollector(this.reply)

        this.collector.on('collect', async (int) => {
            if (this.users.includes(int.user.id)) {
                if (!this.filter(int)) return;
                const id = int.customId;
                if (id === "stop") {
                    await int.update(this.getOpts(true))
                    return this.collector.stop('button');
                }
                else {
                    switch (id) {
                        case 'next':
                            await this.next(int);
                            break;
                        case 'previous':
                            await this.previous(int);
                            break;
                    }
                }
            }
            await int.deferUpdate().catch(() => 0);
        })

        this.collector.on('end', async (_, reason) => {
            if (reason && reason.toLowerCase().includes('delete')) this.deleted = true;
            return await this.end(reason).catch(console.error)
        });
        return this;
    }
}

index.d.ts
import {
    Client,
    InteractionCollector,
    Message,
    ActionRowBuilder,
    Interaction,
    ButtonStyle,
    BaseMessageOptions,
    EmbedBuilder,
    UserResolvable,
    ButtonInteraction
} from "discord.js";

// Types
type ButtonId = "previous" | "stop" | "next" | string
type Embed = EmbedBuilder
type User = UserResolvable
type Context = Message | Interaction

declare class Dispage {
    /** @param {Client} client  */
    constructor(client: Client);
    client: Client<boolean>;
    stopReason: string;
    index: number;
    /** @type {Embed[]} */
    embeds: Embed[];
    message?: Message;
    interaction?: Interaction;
    collector: InteractionCollector<ButtonInteraction>;
    reply: Message;
    ended: boolean;
    started: boolean;
    deleted: boolean;
    duration: number;
    mainStyle: ButtonStyle;
    styles: string[];
    buttons: {
        customId: string;
        emoji: string;
        style: string;
        label: string;
    }[];
    filter: (int: ButtonInteraction) => boolean;
    _hdb: boolean;
    /** @type {Set<string>} */
    _userIDs: Set<string>;
    get users(): string[];
    get ctx(): "MESSAGE" | "INTERACTION";
    get type(): string;
    get endUntill(): number;
    /** @returns {Embed} */
    get currentEmbed(): Embed;
    _getId(user: any): any;
    setMainStyle(style: ButtonStyle): this;
    hideDisabledButtons(should?: boolean): this;
    removeUser(user: User): this;
    addUser(user: User): this;
    setUser(user: User): this;
    addButton(o: object): this;
    removeAllButtons(): this;
    findButtonAndRemove(query: Function): this;
    removeButton(customId: string): this;
    editButton(customId: string, o: object): this;
    getRows(disabled?: boolean): ActionRowBuilder | null;
    setEmbeds(embeds: Embed[]): this;
    addEmbed(embed: Embed): this;
    addEmbeds(embeds: Embed[]): this;
    _fixEmbeds(embeds: Embed | Embed[] | Embed[][]): Embed[];
    setDuration(duration: number): this;
    addDuration(duration: number): this;
    next(int: any): Promise<Message | false>;
    previous(int: any): Promise<Message | false>;
    doesIndexExist(index: number): boolean;
    changeToPage(index: number, int: any): Promise<Message>;
    setIndex(index: number): this;
    end(reason: string): Promise<void | Message>;
    delete(): Promise<Message>;
    getOpts(disabled?: boolean): BaseMessageOptions;
    isMessage(): boolean;
    isInteraction(): boolean;
    canEdit(): boolean;
    isValidCtx(ctx: Context): boolean;
    _createCollector(reply: Message): InteractionCollector<any>;
    checkForErrors(ctx: any): string[];
    start(ctx: Context): Promise<this>;
}
export = Dispage;


cached.d.ts
import {
    Client,
    InteractionCollector,
    UserResolvable,
    MessageEmbed,
    Interaction,
    Message,
    MessageOptions,
    MessageEditOptions,
    MessageActionRow,
    MessageButtonOptions,
    MessageButtonStyle,
    ButtonInteraction
} from "discord.js";

type Context = Discord.Interaction | Discord.Message

export = Dispage;
declare class Dispage {
    setMainStyle(style: MessageButtonStyle): this;
    removeUser(user: UserResolvable): this;
    addUser(user: UserResolvable): this;
    setUser(user: UserResolvable): this;
    addButton(o: MessageButtonOptions): this;
    removeButton(customId: string): this;
    editButton(customId: string, o: object): this;
    getRow(disabled?: boolean): MessageActionRow | null;
    setFooter(func: Function): this;
    setEmbeds(embeds: MessageEmbed[]): this;
    shouldShowDisabledButtons(should?: boolean): this;
    addEmbed(embed: MessageEmbed): this;
    fixEmbeds(embeds: MessageEmbed | MessageEmbed[] | MessageEmbed[][]): MessageEmbed[];
    addEmbeds(embeds: MessageEmbed[]): this;
    setDuration(duration: number): this;
    addDuration(duration: number): this;
    next(): Promise<void> | this;
    previous(): Promise<void> | this;
    setIndex(index: number): this;
    fixEmbedFooters(): void;
    edit(opts: MessageEditOptions): Promise<Message<boolean>>;
    disableButtons(): Promise<Message<boolean>>;
    end(): Promise<void> | Promise<Message<boolean>>;
    delete(): Promise<Message<boolean>>;
    update(): Promise<Message<boolean>>;
    getOpts(disabled?: boolean): MessageOptions;
    canEdit(): boolean;
    isMessage(): boolean;
    isInteraction(): boolean;
    isValidCtx(ctx: Context): boolean;
    checkForErrors(msg: Context): string[];
    start(ctx: Context): Promise<this>;
    /** @deprecated Use .setUser() instead. */
    private setUserID(user: UserResolvable): this;
    private _getId(user: UserResolvable): string;
}


example.js
const Dispage = require('dispage')
const Discord = require('discord.js')
const client = new Discord.Client({
    intents: [
        Discord.GatewayIntentBits.Guilds,
        Discord.GatewayIntentBits.GuildMessages,
        Discord.GatewayIntentBits.MessageContent
    ]
});

client.on('ready', () => console.log('I am ready!'))

client.on('messageCreate', message => {
    if (message.content === "!test") {

        const embeds = [
            'embed 1',
            'embed 2 !!',
            'embed 3 !?!',
            'embed 4 !!!!!!!!?'
        ].map((desc, i) => new Discord.EmbedBuilder()
            .setTitle(`📃 Page #${i + 1}`)
            .setColor('BLURPLE')
            .setDescription(desc))

        new Dispage()
            .setEmbeds(embeds)
            // This style will be given to every button that doesn't have a specific style
            .setMainStyle('Secondary')
            .editButton('previous', { emoji: "824240081409540157", label: "Previous" })
            // Setting a property (here, style) to null will remove it. 👇
            .editButton('stop', { emoji: "961338862259544114", label: null })
            .editButton('next', { emoji: "824240024509874186", label: "Next" })
            // We can edit as many buttons as we want and as many times as we wish
            .editButton('stop', { style : "Success" })
            // You can create your own custom button
            .addButton({
                label: "Go to Message",
                style: "Link",
                url: "https://discord.com/channels/937626764916719626/937633296878293002/978956060972974101"
            })
            // Finally after having setup everything. We can start
            .start(message)
    }
});

client.login('YOUR TOKEN HERE')